exploration/function-composition-part-1.md

```diff
@@ -0,0 +1,974 @@
+what can we assume about the resolved values
+of `$x` and `$y`?
+
+It must be possible to distinguish
+the resolved value of `$x`
+from the resolved value of `$y`,
+as `{{$x} {$y}}` formats to "1.00 1.001"
+while `{{$x} {$x}}` formats to "1.00 1.00".
+
+The question is _when_ the two resolved values
+look the same,
+and when they look different.
+
+Two possible answers:
+
+1. The two resolved values are always different:
+`$x` is bound to a resolved value
+that encodes the value of the `maxFrac` option, 2;
+while for `$y`, this value is 3.
+2. Functions cannot distinguish the two resolved values
+from each other.
+However, the formatter can distinguish the two resolved values
+from each other when formatting a pattern.
+
+How would the spec need to change in order to
+force either answer 1 or answer 2 to be unambiguously true?
+
+Put a different way: does every expression
+have a single resolved value
+in a given formatting context?
+Or does the resolved value of an expression
+depend on the context
+in which the expression appears?
+
+### Custom and built-in functions
+
+An implementation that supports custom functions
+would be expected to define an interface between the message formatter
+and function implementations.
+An implementation is free to choose whether to use the same interface
+for calling built-in functions,
+or to build the implementations of those functions
+directly into the message formatter.
+
+For that reason, this document refers to custom functions
+when discussing the function interface.
+However, in some implementations, the same questions arise
+for built-in functions.
+
+### Formattable and formatted values
+
+In implementing the custom function registry,
+it might be natural to suppose that
+conceptually, the type signature of a function implementation
+is:
+
+```
+Formattable -> FormattedValue
+```
+
+
+The names of these two types are taken from the ICU4C
+implementation. A `Formattable` is a value
+tagged with a type (for example, doubles, integers,
+strings, and pointers to arbitrary objects).
+A `FormattedValue` represents something that
+already contains all the information needed
+to either render it as a string,
+or convert it to a "part" according to a hypothetical
+"format to parts" interface.
+
+It would be tempting to use the existing `Formattable`
+type for the input of a function implementation,
+and use the existing `FormattedValue` type for the output.
+However, functions with this signature
+can't compose with each other.
+
+When the following text refers to `Formattable` or
+`FormattedValue`, it should be taken to refer to
+abstract "input" and "output" types.
+
+### Ambiguous examples
+
+Returning to Example Y1, consider two possible models
+of the runtime behavior of function composition.
+
+**Model 1:**
+
+1. Evaluate `$num` to a value and pass it to the `:number` function,
+   along with named options `{"maxFrac": "2"}`
+2. Bind `$x` to the result, which is an object `X`
```

Comment:
semantics question: I always thought you evaluate, get a value, and then bind to symbols/names. which "binds to" which?
```suggestion
2. Bind the result to `$x`, which is an object `X`
```

Reply:
I've usually seen "bind a name to a value", but I rewrote these steps (and the next set of steps) to hopefully make it clearer, in [981df0f](https://github.com/unicode-org/message-format-wg/pull/753/commits/981df0fc454642f55bf58f227bf9b7fda0926a15)